devito/ir/analysis/basic.py

```diff
@@ -0,0 +1,182 @@
+        if not all(is_integer(i) or isinstance(i, Basic) for i in items):
+            raise TypeError("Illegal Vector element type")
+        return super(Vector, cls).__new__(cls, items)
+
+    def _asvector(func):
+        def wrapper(self, other):
+            if not isinstance(other, Vector):
+                try:
+                    other = Vector(*other)
+                except TypeError:
+                    # Not iterable
+                    other = Vector(*(as_tuple(other)*len(self)))
+            if len(self) != len(other):
+                raise TypeError("Cannot operate with Vectors of different rank")
+            return func(self, other)
+        return wrapper
+
+    @_asvector
+    def __add__(self, other):
+        return Vector(*[i + j for i, j in zip(self, other)])
+
+    @_asvector
+    def __radd__(self, other):
+        return self + other
+
+    @_asvector
+    def __sub__(self, other):
+        return Vector(*[i - j for i, j in zip(self, other)])
+
+    @_asvector
+    def __rsub__(self, other):
+        return self - other
+
+    @_asvector
+    def __eq__(self, other):
+        return super(Vector, self).__eq__(other)
+
+    @_asvector
+    def __ne__(self, other):
+        return super(Vector, self).__ne__(other)
+
+    @_asvector
+    def __lt__(self, other):
+        try:
+            diff = [int(i) for i in self.order(other)]
+        except TypeError:
+            raise TypeError("Cannot compare due to non-comparable index functions")
+        return diff < [0]*self.rank
+
+    @_asvector
+    def __gt__(self, other):
+        return other.__lt__(self)
+
+    @_asvector
+    def __ge__(self, other):
+        return self.__eq__(other) or self.__gt__(other)
+
+    @_asvector
+    def __le__(self, other):
+        return self.__eq__(other) or self.__lt__(other)
+
+    def __getitem__(self, key):
+        ret = super(Vector, self).__getitem__(key)
+        return Vector(*ret) if isinstance(key, slice) else ret
+
+    def __repr__(self):
+        maxlen = max(3, max([len(str(i)) for i in self]))
+        return '\n'.join([('|{:^%d}|' % maxlen).format(str(i)) for i in self])
+
+    @property
+    def rank(self):
+        return len(self)
+
+    @property
+    def sum(self):
+        return sum(self)
+
+    def distance(self, other):
+        """Compute vector distance from ``self`` to ``other``."""
+        return self - other
+
+    def order(self, other):
+        """
+        A reflexive, transitive, and anti-symmetric relation for total ordering.
+
+        Return a tuple of length equal to the Vector ``rank``. The i-th tuple
+        entry, of type int, indicates whether the i-th component of ``self``
+        precedes (< 0), equals (== 0), or succeeds (> 0) the i-th component of
+        ``other``.
+        """
+        return self.distance(other)
+
+
+class IterationInstance(Vector):
+
+    """A representation of the iteration space point accessed by a
+    :class:`Indexed` object."""
```

Comment:
This sounds a little vague. Would a little more detail help understand why this is needed?

Reply:
All the docstrings are already carefully expanded/modified in the future PRs!

reference branch is ``optimize-timeslots-final``. However, it's too premature to copy those docstrings here (due to missing functionalities).

Reply:
Ah ok. Thanks